core/src/main/java/io/substrait/relation/ProtoRelConverter.java

```diff
@@ -0,0 +1,249 @@
+    return Project.builder()
+        .input(input)
+        .remap(optionalRelmap(rel.getCommon()))
+        .expressions(rel.getExpressionsList().stream().map(converter::from).toList())
+        .build();
+  }
+
+  private Aggregate newAggregate(AggregateRel rel) {
+    var input = from(rel.getInput());
+    var converter = new ProtoExpressionConverter(lookup, extensions, input.getRecordType());
+    List<Aggregate.Grouping> groupings = new ArrayList<>(rel.getGroupingsCount());
+    for (var grouping : rel.getGroupingsList()) {
+      groupings.add(
+          Aggregate.Grouping.builder()
+              .expressions(
+                  grouping.getGroupingExpressionsList().stream().map(converter::from).toList())
+              .build());
+    }
+    List<Aggregate.Measure> measures = new ArrayList<>(rel.getMeasuresCount());
+    for (var measure : rel.getMeasuresList()) {
+      var func = measure.getMeasure();
+      measures.add(
+          Aggregate.Measure.builder()
+              .function(
+                  AggregateFunctionInvocation.builder()
+                      .arguments(
+                          measure.getMeasure().getArgsList().stream().map(converter::from).toList())
+                      .declaration(
+                          lookup.getAggregateFunction(func.getFunctionReference(), extensions))
+                      .outputType(FromProto.from(func.getOutputType()))
+                      .aggregationPhase(Expression.AggregationPhase.fromProto(func.getPhase()))
+                      .build())
+              .preMeasureFilter(
+                  Optional.ofNullable(
+                      measure.hasFilter() ? converter.from(measure.getFilter()) : null))
+              .build());
+    }
+    return Aggregate.builder()
+        .input(input)
+        .groupings(groupings)
+        .measures(measures)
+        .remap(optionalRelmap(rel.getCommon()))
+        .build();
+  }
+
+  private Sort newSort(SortRel rel) {
+    var input = from(rel.getInput());
+    var converter = new ProtoExpressionConverter(lookup, extensions, input.getRecordType());
+    return Sort.builder()
+        .input(input)
+        .remap(optionalRelmap(rel.getCommon()))
+        .sortFields(
+            rel.getSortsList().stream()
+                .map(
+                    field ->
+                        Expression.SortField.builder()
+                            .direction(Expression.SortDirection.fromProto(field.getDirection()))
+                            .expr(converter.from(field.getExpr()))
+                            .build())
+                .toList())
+        .build();
+  }
+
+  private Join newJoin(JoinRel rel) {
+    Rel left = from(rel.getLeft());
+    Rel right = from(rel.getRight());
+    Type.Struct leftStruct = left.getRecordType();
+    Type.Struct rightStruct = right.getRecordType();
+    Type.Struct unionedStruct = Type.Struct.builder().from(leftStruct).from(rightStruct).build();
```

Comment:
Need to add more tests for joins with post filters. This may be too simplistic?

Reply:
Does calling from twice concatenate? I wouldn't have expected that.